packages/loader/container-loader/src/connectionManager.ts

```diff
@@ -276,13 +273,6 @@ export class ConnectionManager implements IConnectionManager {
 			  };
 	}
 
-	public shouldJoinWrite(): boolean {
-		// We don't have to wait for ack for topmost NoOps. So subtract those.
-		return (
-			this.clientSequenceNumberObserved < this.clientSequenceNumber - this.localOpsToIgnore
```

Comment:
This code was skipping write mode when there were only no ops. Is this not needed anymore? Will it now connect in write mode if there are only no ops processed?

Reply:
My understanding is that no ops are sent by the Container layer, and Container tracking outstanding ops minus no ops is basically duplicating the dirty signal from runtime.

Reply:
Curtis pointed to me (and I extended it a bit) that old logic might have been incorrect (in a way that is not substantial), like we will connect in "write" mode even though sender (Container Ruintime after rebasing pending ops OR code proposal) might not need / want to send any ops.

That said, I think predicting where we got it wrong and where dirty state might not be fully capturing it (due to bugs) is close to impossible. So I suggested earlier to keep existing logic, but add telemetry that would tell us when isDirty() !== shouldJoinWrite() and try to make sense of data. That will tell us if we could make this change (and what it possibly takes to make it safe).

We could (for starters, not to regress substantially situation) do OR here, i.e. isDirty() || shouldJoinWrite() (it could backfire, but it's probably safer than just removing current logic without getting more data)